lib/natalie/compiler/pass0c.rb

```diff
@@ -16,7 +16,33 @@ def initialize(context)
       end
 
       def process_iter(exp)
-        call_type, call, args, *body = exp
+        _, call, args, *body = exp
+        if call[1]&.sexp_type == :array && body.length == 1
+          # FIXME: This might unwrap/duplicate big blocks of code, if they are
+          #        passed as the function body
+          # FIXME: The implementation of these do not honour name shadowing of
+          #        nested lambdas/iters, yet
+          # FIXME: We do not support re-assignment of any argument passed in yet,
+          #        for that we would need a more sophisticated inlining algorithm
+
+          if !(
+               args.any? do |arg|
+                 body.flatten.any? { |sexp| sexp.is_a?(Sexp) && sexp.sexp_type == :lasgn && sexp[1] == arg }
+               end
+             )
+            _, *values = call[1]
+            case call[2]
+            when :min
+              return process(canary_array_min(exp, values, args, *body))
+            when :max
+              return process(canary_array_max(exp, values, args, *body))
+            when :any?
+              return process(canary_array_any(exp, values, args, *body))
+            when :all?
+              return process(canary_array_all(exp, values, args, *body))
+            end
+          end
+        end
```

Comment:
As this file grows, it's going to be really hard to understand what these various optimizations are doing. At the very least, we should move this code to a method called `optimize_single_value_array_operations` or some such, just so it's easy to talk about and see at-a-glance what the intention is. Otherwise it's just a big block of AST manipulation and my eyes glaze over.